File: tools/PreviewLocomotion/hunt-motion.js
(function(root){
  const clamp=x=>Math.max(0,Math.min(1,x)),ease=x=>{x=clamp(x);return x*x*(3-2*x)};
  const duration=2.6;
  function sample(t){
    if(!Number.isFinite(t))throw Error('Finite time required');
    const amount=ease((t-.25)/.45)*(1-ease((t-2.05)/.55));
    const envelope=ease((t-.7)/.18)*(1-ease((t-1.65)/.25));
    return {amount,headBob:6.3*amount,headRoll:-.07*amount,sway:1.425*Math.sin((t-.75)*Math.PI*5)*envelope,
      label:t<.25?'발견 · 잠깐 집중':t<.7?'앞발 뻗고 가슴 낮추기':t<1.9?'엉덩이 씰룩씰룩':t<2.05?'잠깐 노려보기':t<duration?'자세 풀기':'다시 쉬기'};
  }
  function head(x,y,p){
    const a=p.headRoll,dx=x-43,dy=y-48;
    return [43+dx*Math.cos(a)-dy*Math.sin(a),48+dx*Math.sin(a)+dy*Math.cos(a)+p.headBob];
  }
  function floor(x){
    const points=[[23,79],[43,85],[54.5,78.5],[66,82]];
    for(let i=1;i<points.length;i++)if(x<=points[i][0]){
      const a=points[i-1],b=points[i],u=clamp((x-a[0])/(b[0]-a[0]));return a[1]+(b[1]-a[1])*u;
    }return 82;
  }
  function map(x,y,p){
    const h=head(x,y,p),rear=ease((x-48)/23),contact=1-ease((y-60)/(floor(x)-60));
    // Begin compression across the torso, not just the final short foreleg
    // segment: deeper lowering must not fold the limb mesh onto itself.
    // Forefeet slide forward during entry, then stay planted during the loop.
    // Distribute the reach from the upper foreleg to the wrist. Concentrating
    // eleven pixels of reach in a five-pixel wrist made a hooked S-bend;
    // the longer attachment keeps the capsule-shaped toe joined smoothly.
    const fore=1-ease((x-46)/8),rootY=66+8*ease((x-24)/12),reach=ease((y-rootY)/11);
    // Round the leading shoulder of each pad instead of letting the old
    // outer ankle corner become a pointed fin. Pad centers/soles stay fixed.
    const toeCenter=23+20*ease((x-27)/5),rounding=1-ease((x-toeCenter+7)/7);
    const px=x+((h[0]-x)*(1-rear)+p.sway*rear)*contact-(11-2.2*rounding)*p.amount*fore*reach;
    const py=y+((h[1]-y)*(1-rear)-1.6*p.amount*rear+.22*p.sway*rear)*contact;
    // Add pad volume in destination space, with a monotone vertical warp.
    // Applying it before the large wrist shear could fold the ankle cells.
    const farToe=23-11*p.amount;
    const padFloor=79+6*ease((px-farToe-3)/9),padZone=1-ease((px-44)/9);
    const padLift=1.15*p.amount*padZone*ease((py-padFloor+9)/5)*ease((padFloor-py)/3);
    return [px,py-padLift];
  }
  function createSession(){
    let start=null,releaseAt=null;
    return {update(now,near){
      if(!Number.isFinite(now))throw Error('Finite session time required');
      if(start===null){if(!near)return duration;start=now;}
      const elapsed=Math.max(0,now-start);
      // 1.25 and1.65 are the same settled sway peak (one full period).
      // Keep replaying that middle span, never the crouch entry or recovery.
      // On departure join the source clip's eased tail at the next matching
      // peak, where velocity is zero, instead of snapping the rump to center.
      if(!near&&releaseAt===null)releaseAt=start+1.65+Math.max(0,Math.ceil((elapsed-1.65)/.4-1e-9))*.4;
      if(near&&releaseAt!==null&&now<releaseAt)releaseAt=null;
      if(releaseAt!==null&&now>=releaseAt){
        const age=1.65+now-releaseAt;
        if(age>=duration){start=null;releaseAt=null;return duration;}
        return age;
      }
      return elapsed<1.65?elapsed:1.25+(elapsed-1.25)%.4;
    }};
  }
  const api={sample,map,head,duration,createSession};
  if(typeof module!=='undefined')module.exports=api;else root.huntMotion=api;
})(globalThis);

File: src/Dororong.App/Controls/HuntMotion.cs
using Dororong.Core.Geometry;

namespace Dororong.App.Controls;

internal readonly record struct HuntPose(double Amount, double HeadBob, double HeadRoll, double Sway)
{
    public static HuntPose At(double seconds)
    {
        if (!double.IsFinite(seconds))
            throw new ArgumentOutOfRangeException(nameof(seconds), "Finite time required.");

        var amount = Ease((seconds - 0.25) / 0.45) * (1 - Ease((seconds - 2.05) / 0.55));
        var envelope = Ease((seconds - 0.7) / 0.18) * (1 - Ease((seconds - 1.65) / 0.25));
        return new HuntPose(
            amount,
            6.3 * amount,
            -0.07 * amount,
            1.425 * Math.Sin((seconds - 0.75) * Math.PI * 5) * envelope);
    }

    public PointD Head(double x, double y)
    {
        var dx = x - 43;
        var dy = y - 48;
        return new PointD(
            43 + (dx * Math.Cos(HeadRoll)) - (dy * Math.Sin(HeadRoll)),
            48 + (dx * Math.Sin(HeadRoll)) + (dy * Math.Cos(HeadRoll)) + HeadBob);
    }

    public PointD Map(double x, double y)
    {
        var head = Head(x, y);
        var rear = Ease((x - 48) / 23);
        var contact = 1 - Ease((y - 60) / (Floor(x) - 60));
        var fore = 1 - Ease((x - 46) / 8);
        var rootY = 66 + (8 * Ease((x - 24) / 12));
        var reach = Ease((y - rootY) / 11);
        var toeCenter = 23 + (20 * Ease((x - 27) / 5));
        var rounding = 1 - Ease((x - toeCenter + 7) / 7);
        var px = x + ((head.X - x) * (1 - rear) + (Sway * rear)) * contact
            - (11 - (2.2 * rounding)) * Amount * fore * reach;
        var py = y + ((head.Y - y) * (1 - rear) - (1.6 * Amount * rear) + (0.22 * Sway * rear)) * contact;
        var farToe = 23 - (11 * Amount);
        var padFloor = 79 + (6 * Ease((px - farToe - 3) / 9));
        var padZone = 1 - Ease((px - 44) / 9);
        var padLift = 1.15 * Amount * padZone * Ease((py - padFloor + 9) / 5) * Ease((padFloor - py) / 3);
        return new PointD(px, py - padLift);
    }

    private static double Floor(double x)
    {
        ReadOnlySpan<PointD> points = [new(23, 79), new(43, 85), new(54.5, 78.5), new(66, 82)];
        for (var i = 1; i < points.Length; i++)
        {
            if (x > points[i].X)
                continue;

            var a = points[i - 1];
            var b = points[i];
            var u = Clamp01((x - a.X) / (b.X - a.X));
            return a.Y + ((b.Y - a.Y) * u);
        }

        return 82;
    }

    private static double Ease(double value)
    {
        value = Clamp01(value);
        return value * value * (3 - (2 * value));
    }

    private static double Clamp01(double value) => Math.Max(0, Math.Min(1, value));
}

internal readonly record struct HuntGazePose(double EyeX, double EyeY, double Roll);

internal sealed class HuntGaze
{
    private double _eyeX;
    private double _eyeY;
    private double _headX;
    private double _headY;

    public HuntGazePose Advance(double seconds, PointD? target, bool flip, bool active)
    {
        ValidateDelta(seconds);
        var valid = active && target is { } point && double.IsFinite(point.X) && double.IsFinite(point.Y);
        var x = valid ? Clamp(target!.Value.X / 110) * (flip ? -1 : 1) : 0;
        var y = valid ? Clamp(target!.Value.Y / 85) : 0;
        var eye = 1 - Math.Exp(-seconds / 0.075);
        var head = 1 - Math.Exp(-seconds / 0.23);
        _eyeX += (x - _eyeX) * eye;
        _eyeY += (y - _eyeY) * eye;
        _headX += (x - _headX) * head;
        _headY += (y - _headY) * head;
        return new HuntGazePose(
            0.85 * _eyeX,
            0.65 * _eyeY,
            ((0.4375 * _headX) - (0.5625 * _headY)) * (Math.PI / 9));
    }

    public void Reset() => _eyeX = _eyeY = _headX = _headY = 0;

    private static double Clamp(double value) => Math.Max(-1, Math.Min(1, value));

    private static void ValidateDelta(double seconds)
    {
        if (!double.IsFinite(seconds) || seconds < 0)
            throw new ArgumentOutOfRangeException(nameof(seconds), "Nonnegative finite delta required.");
    }
}
